[DateLoader] more clearly expose 'default_collate' and 'default_convert' to users

NivekT opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #69862

Fixes #69445

cc @SsnL @VitalyFedyunin @ejguan @NivekT

Differential Revision: [D33068792](https://our.internmc.facebook.com/intern/diff/D33068792)

NivekT committed:
[DateLoader] more clearly expose 'default_collate' and 'default_convert' to users

[ghstack-poisoned]

SsnL commented:
Can we not expose functions with no complete docstrings?

SsnL commented:
Also I would much rather have a complete solution (e.g. https://github.com/pytorch/pytorch/issues/33181) than small patches like this. Maybe DataPipe has something for this, but I don't know the details.

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33068792).

ejguan commented:
> Can we not expose functions with no complete docstrings?

Agree! It should be easily achieved https://github.com/pytorch/pytorch/pull/69851#issuecomment-992694879

> Also I would much rather have a complete solution (e.g. #33181) than small patches like this. Maybe DataPipe has something for this, but I don't know the details.

I don't think we would do that much effort because DataPipe would provide the flexibility for users to use custom collate as a callable function and apply it on different datasets.

NivekT committed:
Update on "[DateLoader] more clearly expose 'default_collate' and 'default_convert' to users"


Fixes #69445

cc @SsnL @VitalyFedyunin ejguan @NivekT

Differential Revision: [D33068792](https://our.internmc.facebook.com/intern/diff/D33068792)

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33068792).

NivekT commented on torch/utils/data/_utils/collate.py:
```diff
@@ -51,8 +82,49 @@ def default_convert(data):
 
 def default_collate(batch):
-    r"""Puts each data field into a tensor with outer dimension batch size"""
+    r"""
```
@ejguan Have a look and let me know what you think.

ejguan commented on torch/utils/data/_utils/collate.py:
```diff
@@ -51,8 +82,49 @@ def default_convert(data):
+        * `int` -> :class:`torch.Tensor`
+        * `str` -> `str` (unchanged)
+        * `bytes` -> `bytes` (unchanged)
+        * `Mapping[K, V]` -> `Mapping[K, default_collate(V)]`
+        * `NamedTuple[V]` -> `NamedTuple[default_collate(V)]`
+        * `Sequence[V]` -> `Sequence[default_collate(V)]`
```
`default_collate(V)` can make users confused as single value from a batch. Could we do:
`Mapper[K, V_i]` -> `Mapping[K, default_collate([V_1, V_2, ...])]`
`NamedTuple[V1_i, V2_i]` -> `NamedTuple[default_collate([V1_1, V1_2, ...]), default_collate([V2_1, V2_2, ...])]`
`Sequence[V1_i, V2_i, ...]` -> `Sequence[default_collate([V1_1, V1_2, ...]), default_collate([V2_1, V2_2, ...]), ...]`

ejguan reviewed (approved):
LGTM with a nit comment.

NivekT committed:
Update on "[DateLoader] more clearly expose 'default_collate' and 'default_convert' to users"


Fixes #69445

cc @SsnL @VitalyFedyunin ejguan @NivekT

Differential Revision: [D33068792](https://our.internmc.facebook.com/intern/diff/D33068792)

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33068792).
